File: tests/Web.Tests.Playwright/tests/pages/error.page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

/**
 * Error Page Object Model
 */
export class ErrorPage extends BasePage {
  readonly errorHeading: Locator;
  readonly errorMessage: Locator;
  readonly errorCode: Locator;
  readonly homeLink: Locator;
  readonly backButton: Locator;

  constructor(page: Page) {
    super(page);
    this.errorHeading = page.locator('h1:has-text("Error"), h1:has-text("Not Found"), h1:has-text("404")');
    this.errorMessage = page.locator('.error-message, .error-description, p:below(h1)').first();
    this.errorCode = page.locator('text=/404|500|error/i');
    this.homeLink = page.locator('a[href="/"]');
    this.backButton = page.locator('button:has-text("Back"), a:has-text("Go Back")');
  }

  /**
   * Navigate to a non-existent page to trigger 404
   */
  async gotoNonExistentPage(): Promise<void> {
    await super.goto('/this-page-does-not-exist-' + Date.now());
  }

  /**
   * Get the error heading text
   */
  async getErrorHeadingText(): Promise<string> {
    try {
      return await this.errorHeading.textContent() || '';
    } catch {
      return '';
    }
  }

  /**
   * Check if error page is displayed
   */
  async isErrorPageDisplayed(): Promise<boolean> {
    try {
      return await this.errorHeading.isVisible({ timeout: 5000 }) || 
             await this.errorCode.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Check if home link is visible
   */
  async hasHomeLink(): Promise<boolean> {
    try {
      return await this.homeLink.isVisible({ timeout: 5000 });
    } catch {
      return false;
    }
  }

  /**
   * Click on home link to return to home page
   */
  async clickHomeLink(): Promise<void> {
    await this.homeLink.first().click();
    await this.page.waitForLoadState('networkidle');
  }

  /**
   * Click on back button
   */
  async clickBackButton(): Promise<void> {
    if (await this.backButton.isVisible({ timeout: 2000 })) {
      await this.backButton.click();
      await this.page.waitForLoadState('networkidle');
    }
  }
}

File: e2e/Web.Tests.Playwright/PageObjects/ErrorPage.cs
namespace Web.Tests.Playwright.PageObjects;

/// <summary>
/// Error Page Object Model
/// </summary>
[ExcludeFromCodeCoverage]
public class ErrorPage : BasePage
{
	private readonly ILocator _errorHeading;
	private readonly ILocator _errorMessage;
	private readonly ILocator _errorCode;
	private readonly ILocator _homeLink;
	private readonly ILocator _backButton;

	public ErrorPage(IPage page) : base(page)
	{
		_errorHeading = page.Locator("h1:has-text('Error'), h1:has-text('Not Found'), h1:has-text('404')");
		_errorMessage = page.Locator(".error-message, .error-description, p:below(h1)").First;
		_errorCode = page.Locator("text=/404|500|error/i");
		_homeLink = page.Locator("a[href='/']");
		_backButton = page.Locator("button:has-text('Back'), a:has-text('Go Back')");
	}

	/// <summary>
	/// Navigate to a non-existent page to trigger 404
	/// </summary>
	public async Task GotoNonExistentPageAsync()
	{
		await base.GotoAsync($"/this-page-does-not-exist-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}");
	}

	/// <summary>
	/// Get the error heading text
	/// </summary>
	public async Task<string> GetErrorHeadingTextAsync()
	{
		try
		{
			return await _errorHeading.TextContentAsync() ?? string.Empty;
		}
		catch
		{
			return string.Empty;
		}
	}

	/// <summary>
	/// Check if error page is displayed
	/// </summary>
	[Obsolete]
	public async Task<bool> IsErrorPageDisplayedAsync()
	{
		try
		{
			return await _errorHeading.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 5000 }) ||
						 await _errorCode.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 5000 });
		}
		catch
		{
			return false;
		}
	}

	/// <summary>
	/// Check if home link is visible
	/// </summary>
	[Obsolete]
	public async Task<bool> HasHomeLinkAsync()
	{
		try
		{
			return await _homeLink.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 5000 });
		}
		catch
		{
			return false;
		}
	}

	/// <summary>
	/// Click on home link to return to home page
	/// </summary>
	public async Task ClickHomeLinkAsync()
	{
		await _homeLink.First.ClickAsync();
		await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
	}

	/// <summary>
	/// Click on back button
	/// </summary>
	[Obsolete]
	public async Task ClickBackButtonAsync()
	{
		if (await _backButton.IsVisibleAsync(new LocatorIsVisibleOptions { Timeout = 2000 }))
		{
			await _backButton.ClickAsync();
			await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
		}
	}
}